### PR TITLE
Use MDX to test the README example

### PR DESCRIPTION
**Pull request by CraigFe**
Resolves https://github.com/mirage/irmin/issues/986.

- Requires Dune 2.4.0 for the `(mdx)` stanza. Dune 2.0+ locks are currently broken (soon to be fixed by https://github.com/ocaml/dune/pull/3366). This adds a fake dependency `test/irmin-unix/runtest` → `test/irmin-http/runtest` to ensure that those are serialised.

- Requires unreleased MDX support for labels in HTML comments, so marked as draft until those are released.

### CODE REVIEW

**Comment by CraigFe** [comment]
Following the recent MDX release (thanks @NathanReb!), this can now be considered for merge.

**Comment by CraigFe** [comment]
CI fails due to https://github.com/ocaml/dune/pull/3650, so this is blocked.

**Comment by samoht** [comment]
The issue in dune has been fixed in 2.7.0. Could you rebase your PR and see if that makes things go better?

**Comment by CraigFe** [comment]
Travis isn't updating properly to demonstrate it, but I'm fairly sure that this issue _hasn't_ been fixed in Dune 2.7.0.

In the [linked PR](https://github.com/ocaml/dune/pull/3650), I improved the error message in the case where not all packages listed in the `mdx` stanza are available (such as when running with `-p`). However, the fundamental issue is that `mdx` stanzas can't be attached to a _specific_ package in the way that tests can, so multi-package repositories cannot have `mdx` stanzas that do not work with each individual package individually.

@NathanReb is [aware of the issue](https://github.com/ocaml/dune/pull/3650#issuecomment-664322678) and seems to have plans to fix it, although I don't know if there's an issue anywhere tracking that. Until that's fixed, we're once again blocked on using MDX this way.

File: README.md [hunk 1]
```diff
@@ -90,25 +93,23 @@ let info fmt = Irmin_unix.info ~author fmt
...
+  Store.get t [ "foo"; "bar" ] >|= fun x -> Printf.printf "foo/bar => '%s'\n" x
 
 (* Run the program *)
 let () = Lwt_main.run main
 ```
 
 The example is contained in `examples/readme.ml`. It can be compiled and executed with dune:
 
-```bash
+<!-- $MDX skip -->
```
Comment (samoht) [inline README.md:111]: Why is this skipped?

**Comment by samoht** [comment]
> @NathanReb is aware of the issue and seems to have plans to fix it, although I don't know if there's an issue anywhere tracking that. Until that's fixed, we're once again blocked on using MDX this way.

Please make sure an issue is open to keep track of this issue -- we should make sure this is fixed properly!

File: README.md (hunk 1 above)
Comment (CraigFe) [reply README.md:111]: I think `mdx` has issues with running Dune from inside Dune:

```diff
 |```sh
 |$ dune build examples/readme.exe
-|$ dune exec examples/readme.exe
-|foo/bar => 'testing 123'
+|Error: Multiple rules generated for _build/default/examples/readme.exe:
+|- file present in source tree
+|- examples/dune:2
+|Hint: rm -f examples/readme.exe
+|[1]
+|$ dune exec examples/readme.exe
+|Error: Multiple rules generated for _build/default/examples/readme.exe:
+|- file present in source tree
+|- examples/dune:2
+|Hint: rm -f examples/readme.exe
+|[1]
```

but I may just be missing something :slightly_smiling_face:

Comment (samoht) [reply README.md:111]: that's weird, as it's something that is done in RWO without much trouble. Again please report it upstream so that it's tracked and solved properly :-)

Comment (CraigFe) [reply README.md:111]: Looks to me like RWO is cheating slightly by unsetting `INSIDE_DUNE`: https://github.com/realworldocaml/book/blame/master/book/testing/README.md#L544. I'll discuss this with @NathanReb when he gets back from holiday next week.

Comment (samoht) [reply README.md:111]: Please open an issue so I can follow the discussion too ;-)

**Comment by CraigFe** [comment]
The blocking problem is now tracked by https://github.com/ocaml/dune/issues/3756 in Dune.

**Comment by samoht** [comment]
@voodoos is the feature available in `dune`? Can we use it in irmin? :-)

**Comment by voodoos** [comment]
I was not aware of the request for a `package` field in the MDX stanza when I started working on the new version of the stanza which does feature a generic `deps` field instead of the old `packages` field. (For reference, the RFC is here: https://github.com/ocaml/dune/issues/3955)

However the development of the new stanza was slowed down after we discover an issue with ocamlfind in opam monorepos (that is in fact not related to the MDX stanza).

We recently decided to resume work on it with @NathanReb. I rebased the MDX part and will probably rework the Dune part soon. I guess we can take that opportunity to add this `package` field ?

For now it is planned for milestone 3.0. @NathanReb do you have more input on this ?

**Comment by NathanReb** [comment]
> For now it is planned for milestone 3.0. @NathanReb do you have more input on this ?

Not particularly but being able to attach an MDX stanza to a package is definitely a required feature for multi-opam repositories.
There might be issues for users that want to `#require` a package from within the repo in their MDX blocks as there won't be a generic way to express that dependency (unless `(package ...)` deps have been fixed to work in an opam context) but they will be able to work around it by using the `(libraries ...)` field of the new stanza instead.
To summarize: adding the `(package ...)` field is 100% :+1:

**Comment by emillon** [comment]
FYI, `(package)` support has been merged and will be in dune 2.9 (next week) and dune 3.0.

**Comment by emillon** [comment]
Dune 2.9 is now out with that feature.

File: README.md [hunk 2]
```diff
@@ -153,10 +160,14 @@ which can be executed in the same way.
...
 
-    $ echo "root: ." > irmin.yml
-    $ irmin init
-    $ irmin set foo/bar "testing 123"
-    $ irmin get foo/bar
+```bash
+$ echo "root: ." > irmin.yml
+$ irmin init
+$ irmin set foo/bar "testing 123"
+irmin: [WARNING] Reference refs/heads/main not found.
```
Comment (CraigFe) [inline README.md:167]: This will be fixed by https://github.com/mirage/irmin/pull/1698, which should be merged first. (But this PR can still be reviewed separately.)

**Review by zshipko — APPROVED** [review]
LGTM, this makes updating the README much less error prone!

Seeing this makes me wonder if it would make sense for the tutorial code to live in the main repo so it can automatically be checked as well. I can imagine several reasons *not* to do that, but perhaps at some point in the future.
